sys/kern/mmap.c

```diff
@@ -63,22 +63,31 @@ int do_munmap(vaddr_t addr, size_t length) {
   if (length == 0)
     return EINVAL;
 
-  if (!page_aligned_p(addr) || !page_aligned_p(length))
+  if (!page_aligned_p(start) || !page_aligned_p(length))
     return EINVAL;
 
-  vaddr_t right_boundary = addr + length;
+  vaddr_t end = start + length;
 
   WITH_VM_MAP_LOCK (uspace) {
-    while (addr < right_boundary) {
-      vm_map_entry_t *ent = vm_map_find_entry(uspace, addr);
-      if (!ent)
-        return EINVAL;
+    /* Find first entry affected by unmapping memory. */
+    vm_map_entry_t *ent = vm_map_find_entry(uspace, start);
+    if (!ent)
+      return EINVAL;
 
-      vaddr_t end = vm_map_entry_end(ent);
+    while (vm_map_entry_start(ent) <= start && vm_map_entry_end(ent) >= end) {
```

Comment:
I don't understand this condition. What if the end of the first `vm_map_entry_t` is smaller than `end`? Then we won't delete anything and I think we should.

Reply:
As I understand it we are iterating as long as the intersection of the range of `vm_map_entry_t` and `[start, end]` is nonempty. Wouldn't something like `while (vm_map_entry_start(ent) < end)` suffice?

Reply:
Yes, you are right. I repaired it as you suggested.